agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp

```diff
@@ -1814,6 +1818,76 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
     }
 }
 
+bool backhaul_manager::handle_client_capability_query(ieee1905_1::CmduMessageRx &cmdu_rx,
+                                                      const std::string &src_mac)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received CLIENT_CAPABILITY_QUERY_MESSAGE , mid=" << std::dec << int(mid);
+
+    auto client_info_tlv_r = cmdu_rx.getClass<wfa_map::tlvClientInfo>();
+    if (!client_info_tlv_r) {
+        LOG(ERROR) << "getClass wfa_map::tlvClientInfo failed";
+        return false;
+    }
+
+    //Check if it is an error scenario - if the STA specified in the Client Capability Query message is not associated
+    //with any of the BSS operated by the Multi-AP Agent (an error scenario)
+    bool error_scenario = 1;
+    for (const auto &slave : m_radio_info_map) {
+        auto associated_clients_map = slave.second.associated_clients_map;
+        for (const auto &vap : associated_clients_map) {
+            if (vap.second.find(client_info_tlv_r->client_mac()) != vap.second.end()) {
+                error_scenario = 0;
+                break;
+            }
+        }
+    }
+
+    // send CLIENT_CAPABILITY_REPORT_MESSAGE back to the controller
+    if (!cmdu_tx.create(mid, ieee1905_1::eMessageType::CLIENT_CAPABILITY_REPORT_MESSAGE)) {
+        LOG(ERROR) << "cmdu creation of type CLIENT_CAPABILITY_REPORT_MESSAGE, has failed";
+        return false;
+    }
+
+    auto client_info_tlv_t = cmdu_tx.addClass<wfa_map::tlvClientInfo>();
+    if (!client_info_tlv_t) {
+        LOG(ERROR) << "addClass wfa_map::tlvClientInfo has failed";
+        return false;
+    }
+    client_info_tlv_t->bssid()      = client_info_tlv_r->bssid();
+    client_info_tlv_t->client_mac() = client_info_tlv_r->client_mac();
+
+    auto client_capability_report_tlv = cmdu_tx.addClass<wfa_map::tlvClientCapabilityReport>();
+    if (!client_capability_report_tlv) {
+        LOG(ERROR) << "addClass wfa_map::tlvClientCapabilityReport has failed";
+        return false;
+    }
+
+    //if it is an error scenario, set Success status to 0x01 = Failure and do nothing after it.
+    if (!error_scenario) {
+        client_capability_report_tlv->result_code() = wfa_map::tlvClientCapabilityReport::SUCCESS;
+        LOG(DEBUG) << "Result Code: SUCCESS";
+        //TODO: Add frame body of the most recently received (Re)Association Request frame from this client
```

Comment:
So, this...

We have two options here. Either we send a message to the slave that forwards it to the ap_manager that converts it to a bwl call and generates a response message to the slave that gets forwarded to the backhaul manager that generates the response message (with the same MID). Or, we add the ReAssociation Request frame to the CLIENT_ASSOCIATED_NOTIFICATION and cache it in the backhaul manager. The latter would also require us to resend the notification every time a new reassociation request is received by the AP.

If we choose the first option, we have to take into account a possible race condition: by the time the message reaches hostapd, the client may already have left the BSS.

@vitalybu @morantr @tomereli your opinions?

Note that for this PR it doesn't matter much. If it is cached, we'll just need to update the loop in the beginning that finds the associated client to not just set a flag, but to set (a pointer to) the association request frame.

Reply:
I prefer option (2) - caching. I think it is cleaner.
However, the race condition may occur in both options since the client may have left also by the time the message with the updated ReAssociation frame is received in the BH manager

Reply:
I also prefer the second option.